Implement decomposition for aten.tensor_split.tensor_indices_or_sections

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107251

Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

masnesral committed:
Implement decomposition for aten.tensor_split.tensor_indices_or_sections

Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

[ghstack-poisoned]

masnesral commented on torch/_decomp/decompositions.py:
```diff
@@ -1200,6 +1200,25 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
     return torch.split(self, split_sizes, dim)
 
 
+@aten.tensor_split.tensor_indices_or_sections.py_impl(DispatchKey.CompositeImplicitAutograd)
```
@eellison , I couldn't discern if there's a convention for where this goes in the file, so I stuck it next to a couple other "split" decomps. ¯\_(ツ)_/¯

masnesral commented on torch/_decomp/decompositions.py:
```diff
@@ -1200,6 +1200,25 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
+        sections = tensor_indices_or_sections.item()
+        return self.tensor_split(sections, dim)
+    else:
+        indices = [i.item() for i in tensor_indices_or_sections]
```
@eellison Is this list comprehension going to be inefficient?

eellison replied:
It's true that calling .item() in a loop is an anti-pattern. However, for our case currently, .item() will either return a constant, or will graph break, so it doesnt affect us. This tensor is also guaranteed to be on cpu so the `.item()` perf is not quite as bad as on cuda (although it's still bad).

The alternative to .item() in a loop is `.tolist()`. However, `tolist()` isn't supported for tensor subclasses (aka fake tensor). Maybe that should be an op ?

soliciting other opinions, cc @bdhirsh @ezyang

masnesral committed:
fix lints on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on torch/_decomp/decompositions.py:
```diff
@@ -1200,6 +1200,28 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
+    dim: int = 0,
+) -> List[Tensor]:
+    split_dim = tensor_indices_or_sections.dim()
+    assert split_dim == 1 or split_dim == 0, (
```
nit: f strings will format eagerly, even if the assertion passes. we use `torch._check` to lazily format exception messages

eellison commented on torch/_decomp/decompositions.py:
```diff
@@ -1200,6 +1200,28 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
+    assert split_dim == 1 or split_dim == 0, (
+        "tensor_split expected tensor_indices_or_sections to be a zero-dimensional or "
+        f"one-dimensional tensor, but got a tensor with {split_dim} dims"
+    )
```
nit: check split_dtype is cpu and is kLong ?

eellison reviewed (commented):
sweet, looks good ! I think the .item() calls are fine currently given, but lets wait for one other opinion.

masnesral committed:
address nits on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral committed:
fix lint on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral commented:
@pytorchbot merge

masnesral commented:
@pytorchbot rebase

masnesral committed:
Update on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot rebase

masnesral committed:
Update on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral committed:
Update on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

masnesral committed:
Update on "Implement decomposition for aten.tensor_split.tensor_indices_or_sections"


Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
